### PR TITLE
Add Forward Code Restriction

### PR DESCRIPTION
**Pull request by shemnon**
Add Forward Code Restriction

* Add the forward code restriction and update stack validation to
reflect that change.
* Add some verbiage to the motivation about JITs and AOT
* Add some test cases

### CODE REVIEW

File: EIPS/eip-5450.md [hunk 1]
```diff
@@ -54,6 +55,8 @@ For each instruction:
    3. Check if *target offset* is within the code.
    4. Check if *target offset* is not marked as *data*.
    5. Mark *target offset* as *jump target*.
+5. If the prior instruction was a terminal instruction
```
Comment (gumb0) [inline EIPS/eip-5450.md:58]: I think this should include unconditional jump
```suggestion
5. If the prior instruction was a terminal instruction or RJUMP
```

**Comment by gumb0** [comment]
This seems to be a working solution ~~, and the new restriction for forward referenced jump destinations is required only if we want to do everything in one pass together with instruction validation.~~ No, it's required regardless of instruction validation, it's needed because of linear pass through code without worklist.

The main downside seems to be the additional burden for compilers to order the code in a certain way ~~, and this only for the sake of getting rid of additional code pass at validation.~~

**Comment by gumb0** [comment]
as an example of the new restriction, this code would be invalid:
```
RJUMP tag2
tag1:
RJUMP tag3
tag2:
RJUMP tag1
tag3:
STOP
```

File: EIPS/eip-5450.md [hunk 2]
```diff
@@ -62,12 +65,10 @@ The time and memory complexity if this pass is *O(len(code))*.
...
-For each reachable instruction in the code the operand stack height is recorded.
+For each instruction in the code the operand stack height is recorded.
 The first instruction has the recorded stack height equal to the number of inputs to the function matching the code.
 
-The FIFO queue *worklist* is used for tracking "to be inspected" instructions. Initially, it contains the first instruction.
-
-While *worklist* is not empty, pop an instruction and:
+The instructions are processed in-order until there are no more instructions.
 
 1. Determine the effect the instruction has on the operand stack:
```
Comment (gumb0) [inline EIPS/eip-5450.md:73]: It should also reset current recorded stack height after terminating instructions and RJUMP.

**Comment by shemnon** [comment]
But this could be reorderd as such and preserve the semantics:

```
RJUMP tag2
tag2:
RJUMP tag1
tag1:
RJUMP tag3
tag3:
STOP
```
My assertion being any code can be made forward jumpable.  If the compiler already has dead code removal the forward code references should naturally fall out of that analysis.

**Pull request closed without merging**